File: parent/sorting_algo/quick/lomuto.py
# Lomuto Quick Sort
def quick_sort(arr, left, right):
    if left >= right:
        return

    # randomized pivot
    # pivotIndex = left + rand.Intn(right-left+1)
	# nums[pivotIndex], nums[right] = nums[right], nums[pivotIndex]

    pivot = right
    i = left-1

    for j in range(left, right):
        if arr[j] < arr[pivot]:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    
    i += 1
    arr[i], arr[pivot] = arr[pivot], arr[i]

    quick_sort(arr, left, i-1)
    quick_sort(arr, i+1, right)

    # Optimization: Recurse on the smallest side first. Reduces worst-case stack depth from O(n) to O(logn)
    # if i-left < right-i:
    #     quick_sort(nums, left, i-1)
    #     quick_sort(nums, i+1, right)
    # else:
    #     quick_sort(nums, i+1, right)
    #     quick_sort(nums, left, i-1)


# nums = [2, 1, 8, 9, 7, 4]
nums = [1, 5, 2, 3, 4, 1, 4]
quick_sort(nums, 0, len(nums)-1)

print(nums)

File: parent/other_algos/sorting/quick/lomuto.go
package quick

import "fmt"

func main() {
	arr := []int{1, 5, 2, 3, 4, 1, 4}
	quick_sort(arr, 0, len(arr)-1)
	fmt.Println(arr)
}

func quick_sort(nums []int, left, right int) {
	// Lomuto Quick Sort
	if left >= right {
		return
	}
	pivot := right
	i := left - 1

	for j := left; j < right; j++ {
		if nums[j] < nums[pivot] {
			i++
			nums[i], nums[j] = nums[j], nums[i]
		}
	}

	i++
	nums[i], nums[pivot] = nums[pivot], nums[i]

	quick_sort(nums, left, i-1)
	quick_sort(nums, i+1, right)
}
